esctest/escutil.py

```diff
@@ -45,16 +51,109 @@ def GetWindowTitle():
   esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_TITLE)
   return escio.ReadOSC("l")
 
+def can_query_shell_size():
+  """Determine if the terminal responds to a request for shell-window size vs
+  text-window size.
+
+  xterm has two windows of interest:
+  a) the text-window
+  b) the shell-window (i.e., outer window).
+
+  The shell-window includes the title and other decorations.  Unless those are
+  suppressed (an option in some window managers), it will be larger than the
+  terminal's character-cell window.  In any case, it also has a border which is
+  normally present.
+
+  The window operations report for the window-size in pixels was added to xterm
+  to provide an alternate way to measure the text-window (and indirectly, the
+  font size).  That is the default operation.  Alternatively, the size of the
+  shell window can be returned."""
+  global CAN_QUERY_SHELL_SIZE
+  if CAN_QUERY_SHELL_SIZE < 0:
+    CAN_QUERY_SHELL_SIZE = 0
+    # check for the default operation
+    esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_SIZE_PIXELS)
+    params = escio.ReadCSI("t")
+    if params[0] == 4 and len(params) >= 3:
+      # TODO: compare size_1 with result from text-window size
+      # If it is larger, then that's a special case.
+      size_1 = Size(params[2], params[1])
+      CAN_QUERY_SHELL_SIZE = 1
+      # check for the window-shell operation
+      esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_SIZE_PIXELS, 2)
+      params = escio.ReadCSI("t")
+      if params[0] == 4 and len(params) >= 3:
+        size_2 = Size(params[2], params[1])
+        if size_2.height() > size_1.height() and \
+           size_2.width() > size_1.width():
+          CAN_QUERY_SHELL_SIZE = 2
+  return CAN_QUERY_SHELL_SIZE
+
 def GetWindowSizePixels():
   """Returns a Size giving the window's size in pixels."""
-  esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_SIZE_PIXELS)
+  if can_query_shell_size() == 2:
+    esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_SIZE_PIXELS, 2)
+  else:
+    esccmd.XTERM_WINOPS(esccmd.WINOP_REPORT_WINDOW_SIZE_PIXELS)
   params = escio.ReadCSI("t")
   AssertTrue(params[0] == 4)
   AssertTrue(len(params) >= 3)
   return Size(params[2], params[1])
 
+def GetScreenSizePixels():
+  """Returns a Size giving the screen's size in pixels.
+
+  The "screen" refers to the X display on which xterm is running.
+  Because that does not change as a result of these tests, we
+  cache a value to help with performance."""
+  global SCREEN_SIZE_PIXELS
```

Comment:
iTerm2 doesn't support this since the notion of screen is ambiguous when there are multiple displays. I'm not sure if VTE does or not. How about if this returns a fixed value of 1024x768 if the terminal is not xterm?

Reply:
ok